### PR TITLE
[WIP] Add services file for taskwarrior server service

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Adressing #7729 

This is my approach on a service definition for taskserver. Unfortunately, I'm a bit stuck here, **maybe someone can help me?**

[Documentation for taskserver](http://taskwarrior.org/docs/taskserver/setup.html) and how to install.

---

as @bjornfor suggested, a list of things which need to be covered:
- [x] Certificate generation (once)
- [ ] Key generation (once)
- [ ] User creation (on the taskserver) (declarative)
- [ ] Export information the user needs in his `taskrc` file
  - [ ] Key
  - [ ] User-Certificate
  - [ ] CA-CertificateJJ
  - [ ] Credentials

### CODE REVIEW

File: nixos/modules/services/misc/taskserver.nix [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+with lib;
+
+let
+  cfg = config.services.taskserver;
+in {
+
+  options = {
+    services.taskserver = {
+
+      enable = mkEnableOption "Whether to enable the taskwarrior server.";
```
Comment (jagajaga) [inline nixos/modules/services/misc/taskserver.nix]: Change `Whether to enable the taskwarrior server.` -> `taskwarrior server`

File: nixos/modules/services/misc/taskserver.nix [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+      dataDir = mkOption {
+        default = /var/lib/taskserver/data/";
+        description = "Data directory for taskserver";
```
Comment (jagajaga) [inline nixos/modules/services/misc/taskserver.nix]: Forgot dot at the end :)

File: nixos/modules/services/misc/taskserver.nix [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+      caCert = mkOption {
+        description = "Fully qualified path to the CA certificate. Optional.";
+        type = types.path;
+      };
+
+      ciphers = mkOption {
+        default = "NORMAL";
+        description = ''
+          List of GnuTLS ciphers to use. Default is "NORMAL". See your
```
Comment (jagajaga) [inline nixos/modules/services/misc/taskserver.nix]: Don't mention what is default in the description.

File: nixos/modules/services/misc/taskserver.nix [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+          GnuTLS documentation for full details.
+        '';
+        type = types.string;
+      };
+
+      confirmation = mkOption {
+        default = true;
+        description = ''
+          Determines whether certain commands are confirmed.
+          Defaults to on.
```
Comment (jagajaga) [inline nixos/modules/services/misc/taskserver.nix]: Same here :)

File: nixos/modules/services/misc/taskserver.nix [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            request is handled.
+          '';
+          type = types.path;
+        };
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    jobs.taskserver = {
```
Comment (Lassulus) [inline nixos/modules/services/misc/taskserver.nix]: jobs is deprecated, use systemd.services instead.

File: nixos/modules/services/misc/taskserver.nix [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+  cfg = config.services.taskserver;
+in {
+
+  options = {
+    services.taskserver = {
+
+      enable = mkEnableOption "Taskwarrior server.";
+
+      dataDir = mkOption {
+        default = /var/lib/taskserver/data/";
```
Comment (nbp) [inline nixos/modules/services/misc/taskserver.nix]: Add a `"` before `/var/…`.

File: nixos/modules/services/misc/taskserver.nix [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        '';
+        type = types.path;
+      };
+
+      queueSize = mkOption {
+        default = 10;
+        description = ''
+          Size of the connection backlog.  See 'man listen'.
+        '';
+        type = types.int
```
Comment (nbp) [inline nixos/modules/services/misc/taskserver.nix]: This line missing an ending semi-colon.

File: nixos/modules/services/misc/taskserver.nix [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+      environment = {
+        TASKDDATA = "${cfg.dataDir}";
+      };
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/confd";
+        User = cfg.user;
+      };
+    }
```
Comment (nbp) [inline nixos/modules/services/misc/taskserver.nix]: Add a semi-colon after the closing curly brace.

File: nixos/modules/services/misc/taskserver.nix [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+          type = types.string;
+        };
+
+        port = mkOption {
+          default = 53589;
+          description = ''
+            Portnumber of the Taskserver.
+          '';
+          type = types.int;
+        }
```
Comment (nbp) [inline nixos/modules/services/misc/taskserver.nix]: Add a semi-colon after the closing curly brace.

**Comment by nbp** [comment]
You should register this new file in `nixos/modules/module-list.nix`.

**Comment by matthiasbeyer** [comment]
Thanks for your notes. I fixed these points already, but rebased onto a channel-tracking branch (nixos-unstable tracking branch for me) to test-build the service, but I fail to... dependencies cannot be build.

**Comment by geerds** [comment]
What's the current state? How can we help you?

I see two thing we need to do:
- add taskwarrior to `nixos/modules/module-list.nix`
- changelog in `nixos/doc/manual/release-notes/rl-unstable.xml`
- (optional) write a test for the service

Btw you can find syntax errors very easy with nix-instantiate:
```
nix-instantiate --eval-only --strict --xml nixos/modules/services/misc/taskserver.nix
```

**Comment by matthiasbeyer** [comment]
@geerds Thanks for your comment.

I updated the PR with my latest fixes. 

Here is my workflow:
1) take the patches and rebase them onto channel-unstable
2) Add `services.taskserver.enable = true;` to my configuration
3) Try to build the system with `nixos-rebuild build-vm -I nixpkgs=/my/nixpkgs/repo`

Points I have to do:

- [ ] Fix the build. I cannot build because depdenencies fail to build (I don't know why, though)
- [ ] Tests: I don't even know how to write tests (I didn't even know there is a way to test services...)
- [ ] Tests: I'm not really familiar with the whole certificate stuff. I guess we should add tests for taskwarrior with ca-cert stuff, but I guess I cannot do this on my own, sorry!

And yes, the changelog... I will do this as well.

**Comment by matthiasbeyer** [comment]
Rebased onto latest master and updated (there was a conflict in the ids.nix file which I solved).

**Comment by matthiasbeyer** [comment]
nah, accidentially clicked on the wrong button, sorry.

**Comment by matthiasbeyer** [comment]
Can _someone_ please test-build this? I fail hardly, I cannot rebase onto lastest channel-unstable because my system does not build with it, I cannot rebase on an older version because I'd need to change my whole systems configuration for it... waaaah!

**Comment by geerds** [comment]
No problem. I'll try to build it on my machine but it will take one or two days

On May 23, 2015 7:35:13 PM CEST, Matthias Beyer <notifications@github.com> wrote:
>Can _someone_ please test-build this? I fail hardly, I cannot rebase
>onto lastest channel-unstable because my system does not build with it,
>I cannot rebase on an older version because I'd need to change my whole
>systems configuration for it... waaaah!
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/7771#issuecomment-104928858

-- 
  Jascha Geerds
  jg@ekby.de

**Comment by matthiasbeyer** [comment]
> No problem. I'll try to build it on my machine but it will take one or two days

Would be nice. But why would that take so long? Or do you mean your build server or something like that?

**Comment by geerds** [comment]
I'm currently quite busy, that's it :-)

**Comment by matthiasbeyer** [comment]
@geerds Any updates?

**Comment by matthiasbeyer** [comment]
Oh yeah, I was finally able to build my system, so I was now able to build my configuration with this service. Unfortunately it fails to start and I'd really like to find out why, but I guess I'm a bit lost here, as I don't have a clue how to debug this stuff.

Can someone please review my `systemd.services.taskwarrior` set and tell me what I must do differently?

Also I'm insecure with the whole certificate stuff taskserver needs, so if someone could give me an introduction how to set it up correctly, I'd highly appreciate that!

**Comment by lethalman** [comment]
@matthiasbeyer start by looking into journalctl

**Comment by matthiasbeyer** [comment]
@lethalman I found that the service failed to start, because of `preStart` failed with `no such process` or something like that.

**Comment by lethalman** [comment]
Probably preStart has no permission to create that dir. Set `PermissionsStartOnly = true;`

**Comment by matthiasbeyer** [comment]
Hm, issue remains:

    building Nix...
    building the system configuration...
    activating the configuration...
    setting up /etc...
    warning: not applying UID change of user ‘nscd’
    warning: not applying UID change of user ‘ntp’
    warning: not applying UID change of user ‘sshd’
    warning: the following units failed: taskserver.service

    ● taskserver.service - taskserver Service.
       Loaded: loaded (/nix/store/klzlwwkjkb6aki8wclbf7zrlaz4nlr31-unit-taskserver.service/taskserver.service)
       Active: failed (Result: exit-code) since Do 2015-06-04 17:23:00 CEST; 31ms ago
      Process: 14274 ExecStartPre=/nix/store/q9dql36l9ny58snw6v0bz92ds9f614iy-unit-script/bin/taskserver-pre-start (code=exited, status=217/USER)

    Jun 04 17:23:00 lucy systemd[14274]: Failed at step USER spawning /nix/store/q9dql36l9ny58snw6v0bz92ds9f614iy-unit-script/bin/taskserver-pre-start: No such process
    Jun 04 17:23:00 lucy systemd[1]: taskserver.service: control process exited, code=exited status=217
    Jun 04 17:23:00 lucy systemd[1]: Failed to start taskserver Service..
    Jun 04 17:23:00 lucy systemd[1]: Unit taskserver.service entered failed state.
    Jun 04 17:23:00 lucy systemd[1]: taskserver.service failed.
    warning: error(s) occured while switching to the new configuration

**Comment by matthiasbeyer** [comment]
Rebased onto latest master.

**Comment by geerds** [comment]
I'm working on it: https://github.com/geerds/nixpkgs/tree/taskserver

**Comment by matthiasbeyer** [comment]
@geerds Thank you for importing my patches! *dances*

**Comment by geerds** [comment]
Sure! :smiley:

**Comment by matthiasbeyer** [comment]
@geerds Do you have something working (I'm trying out now)? If it works, I would apply your patches onto my PR and squash things up...

**Comment by matthiasbeyer** [comment]
Well the taskserver builds but does not start.

---

Another question: The task executeable wants authorization and certificates in the RC file. Do I actually need this? I'm not sure how to handle this whole stuff.

**Comment by geerds** [comment]
It's not working yet. I'll let you know when I've a working version.

Yeah this whole authorization/certificate stuff is the next point on my to do list

**Comment by matthiasbeyer** [comment]
@geerds something happened by now?

**Comment by matthiasbeyer** [comment]
Rebased. Still WIP.

**Comment by matthiasbeyer** [comment]
Somehow I cannot manage to test build my system with taskserver enabled. It always tells me `services.taskserver` does not exist.

**Comment by matthiasbeyer** [comment]
@geerds Are you alive?

**Comment by matthiasbeyer** [comment]
Some progress, yay! Not tested yet.

**Comment by matthiasbeyer** [comment]
Okay, some bugs are fixed, but maybe someone can help me with the error message:

```
[...]
while evaluating the attribute ‘config.preStart’ at /home/m/archive/development/contrib/nixpkgs/nixos/modules/services/misc/taskserver.nix:193:7:
getting attributes of path ‘/tmp/taskd.log’: No such file or directory
```

I don't see my error.

**Comment by matthiasbeyer** [comment]
So well, this is strange:

I have `services.taskserver.enable = true;` in my `configuration.nix` but I cannot `nixos-rebuild build -I $NIXPKGS_CLONE`, it gives me (with `--show-trace`):

```
building Nix...
building the system configuration...
error: while evaluating the attribute ‘buildCommand’ of the derivation ‘nixos-system-lucy-16.03.git.be77030’ at /home/m/archive/development/contrib/nixpkgs/nixos/modules/system/activation/top-level.nix:102:7:
while evaluating the attribute ‘sources’ of the derivation ‘etc’ at /home/m/archive/development/contrib/nixpkgs/nixos/modules/system/etc/etc.nix:12:5:
while evaluating anonymous function at /home/m/archive/development/contrib/nixpkgs/nixos/modules/system/etc/etc.nix:20:20, called from undefined position:
while evaluating the attribute ‘source’ at /home/m/archive/development/contrib/nixpkgs/lib/attrsets.nix:154:44:
while evaluating anonymous function at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:74:45, called from /home/m/archive/development/contrib/nixpkgs/lib/attrsets.nix:154:52:
while evaluating the attribute ‘value’ at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:287:9:
while evaluating the option `environment.etc.crontab.source':
while evaluating the attribute ‘mergedValue’ at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:314:5:
while evaluating anonymous function at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:314:32, called from /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:314:19:
while evaluating ‘check’ at /home/m/archive/development/contrib/nixpkgs/lib/types.nix:106:15, called from /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:315:10:
while evaluating the attribute ‘allFiles’ of the derivation ‘crontabs’ at /home/m/archive/development/contrib/nixpkgs/pkgs/build-support/trivial-builders.nix:10:14:
while evaluating the attribute ‘text’ of the derivation ‘system-crontab’ at /home/m/archive/development/contrib/nixpkgs/pkgs/build-support/trivial-builders.nix:10:14:
while evaluating the attribute ‘pkgs’ of the derivation ‘system-path’ at /home/m/archive/development/contrib/nixpkgs/pkgs/build-support/trivial-builders.nix:10:14:
while evaluating the attribute ‘environment.systemPackages’ at /home/m/archive/development/contrib/nixpkgs/lib/attrsets.nix:154:44:
while evaluating anonymous function at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:74:45, called from /home/m/archive/development/contrib/nixpkgs/lib/attrsets.nix:154:52:
while evaluating the attribute ‘value’ at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:287:9:
while evaluating the option `environment.systemPackages':
while evaluating the attribute ‘isDefined’ at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:319:5:
while evaluating ‘filterOverrides’ at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:391:21, called from /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:303:18:
while evaluating ‘concatMap’ at /home/m/archive/development/contrib/nixpkgs/lib/lists.nix:62:18, called from /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:397:8:
while evaluating ‘concatMap’ at /home/m/archive/development/contrib/nixpkgs/lib/lists.nix:62:18, called from /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:298:17:
while evaluating anonymous function at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:298:28, called from undefined position:
while evaluating ‘dischargeProperties’ at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:361:25, called from /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:299:62:
while evaluating the attribute ‘value’ at /home/m/archive/development/contrib/nixpkgs/lib/modules.nix:199:48:
while evaluating the attribute ‘config.systemPackages’ at /home/m/config/nixos/lucy.nix:74:3:
while evaluating anonymous function at /nix/store/p0vssf88ya6knw7bz03ijd1bivrz5vbi-nixos/pkgs/networkingPackages.nix:10:1, called from /home/m/config/nixos/lucy.nix:81:19:
while evaluating ‘wlanPkgs’ at /nix/store/p0vssf88ya6knw7bz03ijd1bivrz5vbi-nixos/pkgs/networkingPackages.nix:3:14, called from /nix/store/p0vssf88ya6knw7bz03ijd1bivrz5vbi-nixos/pkgs/networkingPackages.nix:24:7:
while evaluating the attribute ‘config.networking.wireless.enable’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/attrsets.nix:154:44:
while evaluating anonymous function at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:74:45, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/attrsets.nix:154:52:
while evaluating the attribute ‘value’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:287:9:
while evaluating the option `networking.wireless.enable':
while evaluating the attribute ‘isDefined’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:319:5:
while evaluating ‘filterOverrides’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:391:21, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:303:18:
while evaluating ‘concatMap’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/lists.nix:62:18, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:397:8:
while evaluating ‘concatMap’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/lists.nix:62:18, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:298:17:
while evaluating anonymous function at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:298:28, called from undefined position:
while evaluating ‘dischargeProperties’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:361:25, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:299:62:
while evaluating the attribute ‘condition’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:437:14:
while evaluating the attribute ‘condition’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:437:14:
while evaluating the attribute ‘condition’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:437:14:
while evaluating the attribute ‘condition’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:437:14:
while evaluating the attribute ‘enable’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/attrsets.nix:154:44:
while evaluating anonymous function at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:74:45, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/attrsets.nix:154:52:
while evaluating the attribute ‘value’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:287:9:
while evaluating the option `networking.networkmanager.enable':
while evaluating the attribute ‘isDefined’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:319:5:
while evaluating ‘filterOverrides’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:391:21, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:303:18:
while evaluating ‘concatMap’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/lists.nix:62:18, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:397:8:
while evaluating ‘concatMap’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/lists.nix:62:18, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:298:17:
while evaluating anonymous function at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:298:28, called from undefined position:
while evaluating ‘dischargeProperties’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:361:25, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:299:62:
while evaluating the attribute ‘condition’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:437:14:
while evaluating the attribute ‘condition’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:437:14:
while evaluating the attribute ‘condition’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:437:14:
while evaluating the attribute ‘condition’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:437:14:
while evaluating the attribute ‘services.xserver.desktopManager.gnome3’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/attrsets.nix:154:44:
while evaluating anonymous function at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:74:45, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/attrsets.nix:154:52:
while evaluating ‘yieldConfig’ at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:73:29, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:76:16:
while evaluating anonymous function at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:79:24, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:79:11:
while evaluating anonymous function at /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:80:26, called from /nix/store/5jkqr0fg43yim48n05bs09nrrmrvigsq-nixos-16.03pre69426.2a574cb/nixos/lib/modules.nix:80:13:
The option `services.taskserver' defined in `/nix/store/p0vssf88ya6knw7bz03ijd1bivrz5vbi-nixos/services/taskwarrior.nix' does not exist.
```

And I don't know how to debug my `configuration.nix` with `nix-repl` - can someone help me?

**Comment by matthiasbeyer** [comment]
@aszlig Please comment on the diff in the PR the next time. :smiley:

**Comment by aszlig** [comment]
Regarding tests: You could use a snakeoil certificate for that, either a static one or generated [like this](https://github.com/headcounter/deployment/blob/32592ad693499051e8e005c062307e088f407f2f/ssl/snakeoil.nix).

**Comment by aszlig** [comment]
@matthiasbeyer: I did :smiley_cat:

**Comment by matthiasbeyer** [comment]
@aszlig I don't know how to integrate this cert. PRs welcome. And no, you didn't. You commented on the commits.

**Comment by jagajaga** [comment]
@matthiasbeyer https://github.com/matthiasbeyer/nixpkgs/commit/62365596e7393c14bc0248513b55be2294c4015d#commitcomment-13456146

**Comment by aszlig** [comment]
(It makes sense btw. to use the quotes not even on the `if` but also on subsequent lines.

**Comment by aszlig** [comment]
I'll write a VM test for it.

**Comment by matthiasbeyer** [comment]
@jagajaga You're telling me that he did, but the link already says "commitcomment" ... what? Anyways, not the point here, is it? :smile: 

@aszlig Will fix.

**Comment by aszlig** [comment]
`Failed at step USER spawning /nix/store/a28ndnpb1l4z7p2mp7kn7via1j8bbr62-unit-script/bin/taskserver-pre-start: No such process`

That's because the NixOS module doesn't create a `taskd` user.

**Comment by matthiasbeyer** [comment]
Is it enough to

```
users.extraUsers = singleton
  {
    name = cfg.user;
    uid = config.ids.uids.taskd;
    description = "taskd user";
  };
```

?

**Comment by aszlig** [comment]
@matthiasbeyer: Well, you'll get a PR soon :-) Because I'd also add a group.

**Comment by matthiasbeyer** [comment]
@aszlig Wonderful!

**Comment by matthiasbeyer** [comment]
@aszlig Wanna include your vm tests as well?

**Comment by aszlig** [comment]
@matthiasbeyer: I have a few more fixes, will make a separate PR for them.

**Comment by matthiasbeyer** [comment]
I updated this, though I don't think I have time to get this working soon. Help would still be appreciated (@geerds, @aszlig)!

I don't even know how to test this properly... so, well... I cannot do that much, really!

**Comment by matthiasbeyer** [comment]
TIL you can `nix-build` the test.

**Comment by matthiasbeyer** [comment]
The problem here is that the package tarball of the taskwarrior-server package ships some scripts for generating the certificates and keys, though these are not available in the service. Maybe we should alter the taskserver package, so we have multiple outputs there.

**Comment by matthiasbeyer** [comment]
So as far as my VM test works now, I can tell that there are certs and keys beeing generated. I'm not sure about user accounts on the taskserver, though. Maybe @aszlig can tell?

Also: I have to configure my local taskwarrior with the keys etc from the server - how'd I do that with a taskserver service?

**Comment by bjornfor** [comment]
@matthiasbeyer: I see there is a todo list in one of the comments in this thread, but I don't know if it's up to date. Maybe you can have a todo list in the PR cover message? I think that would make it easier to know what issues are left to be solved.

(I once looked briefly at packaging taskserver myself, so I have interest in it. But it's not high on my todo list.)

**Comment by matthiasbeyer** [comment]
@bjornfor I'm not even sure what are the todos in this PR, but I will try to get a short list together.

**Comment by jeaye** [comment]
Any update on this? I was just looking to get a task server setup for my own use and would love to ride on all the coattails I can find.

**Comment by matthiasbeyer** [comment]
@jeaye I need this as well for my personal need, though I'm stuck here. I don't know how to proceed. Feel free to re-use the commits and get something working, feel free to ping me if you have questions (though I'm not sure I can help every time).

**Comment by matthiasbeyer** [comment]
As I cannot get this to work, I will close this PR now. Running `taskd` locally as user works for me and (get-shit-done-ish) that must be enough for me.

Feel free to re-use any commits you want.

**Pull request closed without merging**